src/utils/promiseUtils.ts

```diff
@@ -45,8 +45,37 @@ export class CancellationPromiseSource extends vscode.CancellationTokenSource {
     public readonly promise: Promise<never>;
 
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    public constructor(errorConstructor?: new (...args: any[]) => Error, ...args: any[]) {
+    public constructor(errorConstructor?: new (...args: any[]) => Error, ...args: unknown[]) {
         super();
         this.promise = getCancelPromise(this.token, errorConstructor, args);
     }
 }
+
+export class TimeoutPromiseSource implements vscode.Disposable {
+    private timeoutTimer: NodeJS.Timeout | undefined;
+    private readonly cps: CancellationPromiseSource;
+
+    public constructor(private readonly timeoutMs: number, private readonly context?: IActionContext) {
+        this.cps = new CancellationPromiseSource(Error, localize('vscode-docker.utils.promiseUtils.timeout', 'Request timed out.'));
+    }
+
+    public get promise(): Promise<never> {
+        this.timeoutTimer = setTimeout(() => {
+            if (this.context) {
```

Comment:
Consider having part about suppressing issue reporting in the `callDockerode` code and passing a lambda to the `TimeoutPromiseSource` constructor (instead of `IActionContext`). This will make `TimeoutPromiseSource` more reusable

Reply:
That's fair. Should we generalize and call it just "callback", something that gets called before the cancel?

Reply:
Yes, or "onTimeout", or something like that